### ISSUE TITLE
Gradient of clamp is nan for inf inputs

### ISSUE THREAD

**Issue by arvidfm** [pytorch/pytorch#10729]
## Issue description

The gradient of `torch.clamp` when supplied with `inf` values is `nan`, even when the `max` parameter is specified with a finite value. Normally one would expect the gradient to be `0` for all values larger than `max`, including for `inf`.

## Code example

I'm trying to implement the following piecewise function: `exp(x)` for `x < 0`; `1` otherwise. A straightforward way to do this would be to use `torch.exp(x).clamp(max=1)`, but this does not work as expected:

```py
>>> x = torch.FloatTensor([-10, -5, 0, 5, 10, 50, 60, 70, 80, 90, 100])
>>> x.requires_grad = True
>>> y = torch.exp(x)
>>> y
tensor([4.5400e-05, 6.7379e-03, 1.0000e+00, 1.4841e+02, 2.2026e+04, 5.1847e+21,
        1.1420e+26, 2.5154e+30, 5.5406e+34,        inf,        inf],
       grad_fn=<ExpBackward>)
>>> z = y.clamp(max=1)
>>> z
tensor([4.5400e-05, 6.7379e-03, 1.0000e+00, 1.0000e+00, 1.0000e+00, 1.0000e+00,
        1.0000e+00, 1.0000e+00, 1.0000e+00, 1.0000e+00, 1.0000e+00],
       grad_fn=<ClampMaxBackward>)
>>> z.sum().backward()
>>> x.grad
tensor([0.0000, 0.0067, 1.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000, 0.0000,
           nan,    nan])
```

## System Info

PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: 8.0.61

OS: Arch Linux
GCC version: (GCC) 8.1.1 20180531
CMake version: Could not collect

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 9.2.148
GPU models and configuration: GPU 0: GeForce GT 730
Nvidia driver version: 396.24
cuDNN version: Probably one of the following:
/usr/local/R2017a/bin/glnxa64/libcudnn.so.5.1.5

Versions of relevant libraries:
[pip] numpy (1.14.2)
[pip] numpydoc (0.8.0)
[pip] torch (0.4.1)
[pip] torchvision (0.2.1)
[conda] cuda80                    1.0                           0    soumith
[conda] pytorch                   0.4.1           py36_cuda8.0.61_cudnn7.1.2_1  [cuda80]  pytorch
[conda] torchvision               0.2.1                    py36_1    pytorch

cc @ezyang @albanD @zou3519 @gqchen @pearu @nikitaved @soulitzer

**Comment by sytrus-in-github** [comment]
I guess this is (unfortunately) a normal behavior for auto differentiation process: you have a `0 * inf` situation when doing backprop.  Basically you have `dz/dx = dz/dy * dy/dx`and in this case `dz/dy = 0` cause by clamping and `dy/dx = y = exp(big_number_x) = inf`

In your case I would recommend do clamping before `exp` to avoid `inf` at the first place.

**Comment by arvidfm** [comment]
> In your case I would recommend do clamping before exp to avoid inf at the first place.

Yes, this is how I ended up solving it, but I still think that it would be good to investigate if this behaviour can be fixed, since it's rather unintuitive, especially in the case of `where`.

**Comment by albanD** [comment]
For the `torch.where` example, it is definitely going to be hard to fix as by the time when you compute the backward of the exp, you cannot know if the 0 is a real 0 or from undefined gradients.

For the clamp case, it boils down to the question of: Should the backward pass propagate inf/nan gradients flowing back or try to remove them as much as possible.
I think that in most cases, inf/nan gradients are a problem and so we don't want to silently remove them so that the user is aware of them and can handle them properly (by avoiding them).

**Comment by george-qi** [comment]
Hi all, just wanted to surface some attention for MaskedTensor, a tensor extension library that we just launched the prototype for along PyTorch 1.11. It should help solve many existing problems related to masked semantics, distinguishing between 0 and undefined gradients, etc. 

e.g. for this example:

```
x = torch.tensor([-10., -5, 0, 5, 10, 50, 60, 70, 80, 90, 100], requires_grad=True)
mask = x < 0
mx = masked_tensor(x, mask, requires_grad=True)
my = masked_tensor(torch.ones_like(x), ~mask, requires_grad=True)
y = torch.where(mask, torch.exp(mx), my)
s = y.sum()
s.backward()

print("mx.grad: ", mx.grad)
mx.grad:  masked_tensor(
  [  0.0000,   0.0067,       --,       --,       --,       --,       --,       --,       --,       --,       --]
)
```

Note that the gradient is only provided to selected subset. This effectively changes the gradient of `where` to mask out elements instead of setting them to zero

For more details, you can check out the [website](pytorch.org/maskedtensor) or [Github](github.com/pytorch/maskedtensor). We're actively looking for new users, so if you're interested in using it or have any suggestions on things we should build out, please reach out to me or @cpuhrsch. Thanks!!

**Comment by soulitzer** [comment]
Duplicate of https://github.com/pytorch/pytorch/issues/4132
